author		Francois Mazard <f-mazard@ti.com> 
 		Mon, 18 Apr 2011 08:27:12 +0000 (10:27 +0200) 
committer	Francois Mazard <f-mazard@ti.com> 
 		Mon, 18 Apr 2011 08:51:25 +0000 (10:51 +0200) 
commit		fcd6749d1ef73d3bc2f2a646a87f55b0f5717a43 
tree		3bcd04b80ebdc1cb207fb61dff5d6098d09ff7aa tree 
parent		166d01b32c95b7674b23af3aec82664a3d51002b commit | diff 

ASoC: ABE: Use the right threshold for Modem McBSP frontend

Modem port (stereo), STEREO_RSHIFTED_16 format is used in ABE side,
ATC uses an ITER factor of 2. McBSP threshold should match that value.

Change-Id: Ibd9ec493628c4c0a18ea083bbe4690c96da60218
Signed-off-by: Francois Mazard <f-mazard@ti.com>


Index: kernel/sound/soc/omap/sdp4430.c
===================================================================
--- kernel.orig/sound/soc/omap/sdp4430.c	2011-04-18 15:44:24.177108323 +0900
+++ kernel/sound/soc/omap/sdp4430.c	2011-04-18 19:09:16.909648515 +0900
@@ -76,15 +76,16 @@
 }
 
 static int sdp4430_modem_mcbsp_configure(struct snd_pcm_substream *substream,
-	int flag)
+				struct snd_pcm_hw_params *params, int flag)
 {
 	int ret = 0;
 	struct snd_soc_pcm_runtime *rtd = substream->private_data;
 	struct snd_pcm_substream *modem_substream[2];
 	struct snd_soc_pcm_runtime *modem_rtd;
+	int channels;
 
 	if (flag) {
-		if (!mcbsp_cfg) {
+
 			modem_substream[substream->stream] =
 				snd_soc_get_dai_substream(rtd->card,
 							OMAP_ABE_BE_MM_EXT1,
@@ -95,6 +96,7 @@
 			modem_rtd =
 				modem_substream[substream->stream]->private_data;
 
+		if (!mcbsp_cfg) {
 			/* Set cpu DAI configuration */
 			ret = snd_soc_dai_set_fmt(modem_rtd->cpu_dai,
 					  SND_SOC_DAIFMT_I2S |
@@ -108,6 +110,18 @@
 				mcbsp_cfg = 1;
 			}
 		}
+
+		if (params != NULL) {
+			/* Configure McBSP internal buffer usage */
+			/* this need to be done for playback and/or record */
+			channels = params_channels(params);
+			if (substream->stream == SNDRV_PCM_STREAM_PLAYBACK)
+				omap_mcbsp_set_rx_threshold(
+					modem_rtd->cpu_dai->id, channels);
+			else
+				omap_mcbsp_set_tx_threshold(
+					modem_rtd->cpu_dai->id, channels);
+		}
 	} else {
 		mcbsp_cfg = 0;
 	}
@@ -142,7 +156,7 @@
 
 	if (rtd->current_fe == ABE_FRONTEND_DAI_MODEM) {
 		/* set Modem McBSP configuration  */
-		ret = sdp4430_modem_mcbsp_configure(substream, 1);
+		ret = sdp4430_modem_mcbsp_configure(substream, params, 1);
 	}
 
 	return ret;
@@ -155,7 +169,7 @@
 
 	if (rtd->current_fe == ABE_FRONTEND_DAI_MODEM) {
 		/* freed Modem McBSP configuration */
-		ret = sdp4430_modem_mcbsp_configure(substream, 0);
+		ret = sdp4430_modem_mcbsp_configure(substream, NULL, 0);
 	}
 
 	return ret;
@@ -241,7 +255,7 @@
 
 	if (rtd->current_fe == ABE_FRONTEND_DAI_MODEM) {
 		/* set Modem McBSP configuration  */
-		ret = sdp4430_modem_mcbsp_configure(substream, 1);
+		ret = sdp4430_modem_mcbsp_configure(substream, params, 1);
 	}
 
 	return ret;
@@ -254,7 +268,7 @@
 
 	if (rtd->current_fe == ABE_FRONTEND_DAI_MODEM) {
 		/* freed Modem McBSP configuration */
-		ret = sdp4430_modem_mcbsp_configure(substream, 0);
+		ret = sdp4430_modem_mcbsp_configure(substream, NULL, 0);
 	}
 
 	return ret;
